### PR TITLE
[clang-doc] add support for block commands in clang-doc html output

### LINKED ISSUE #96819: [clang-doc] Clang-Doc not picking up comments for methods in html output
Even in the basic e2e test we can see that it doesn't pick up the comments for the source

see:
https://github.com/llvm/llvm-project/blob/main/clang-tools-extra/test/clang-doc/Inputs/basic-project/include/Calculator.h
https://github.com/llvm/llvm-project/blob/main/clang-tools-extra/test/clang-doc/basic-project.test

I suspect its something wrong with the HTML Generator

### PR DESCRIPTION
**Pull request by PeterChou1**
Fixes https://github.com/llvm/llvm-project/issues/96819

currently the html output for clang-doc does not include block command comments this patch modifies the html generator to generate its output

### CODE REVIEW

File: clang-tools-extra/unittests/clang-doc/HTMLGeneratorTest.cpp [hunk 1]
```diff
@@ -197,7 +197,9 @@ TEST(HTMLGeneratorTest, emitRecordHTML) {
     </p>
     <h2 id="Members">Members</h2>
     <ul>
-      <li>private int X</li>
+      <li>
+        <div>private int X</div>
```
Comment (ilovepi) [inline clang-tools-extra/unittests/clang-doc/HTMLGeneratorTest.cpp:201]: Should these be `private int X`, like this or should we have separate lists for `public`, `protected` and `private`?

File: clang-tools-extra/clang-doc/HTMLGenerator.cpp [hunk 2]
```diff
@@ -418,9 +419,13 @@ genRecordMembersBlock(const llvm::SmallVector<MemberTypeInfo, 4> &Members,
...
-    LIBody->Children.emplace_back(std::make_unique<TextNode>(Access));
-    LIBody->Children.emplace_back(genReference(M.Type, ParentInfoDir));
-    LIBody->Children.emplace_back(std::make_unique<TextNode>(" " + M.Name));
+    auto MemberDecl = std::make_unique<TagNode>(HTMLTag::TAG_DIV);
+    MemberDecl->Children.emplace_back(std::make_unique<TextNode>(Access));
+    MemberDecl->Children.emplace_back(genReference(M.Type, ParentInfoDir));
+    MemberDecl->Children.emplace_back(std::make_unique<TextNode>(" " + M.Name));
+    if (!M.Description.empty())
+      LIBody->Children.emplace_back(genHTML(M.Description));
+    LIBody->Children.emplace_back(std::move(MemberDecl));
```
Comment (ilovepi) [inline clang-tools-extra/clang-doc/HTMLGenerator.cpp:428]: Is this required for Block commands? This seems like a separate change...

Comment (petrhosek) [reply clang-tools-extra/clang-doc/HTMLGenerator.cpp:428]: I agree that this seems like separate change that should be its own PR.

File: clang-tools-extra/unittests/clang-doc/HTMLGeneratorTest.cpp (hunk 1 above)
Comment (petrhosek) [reply clang-tools-extra/unittests/clang-doc/HTMLGeneratorTest.cpp:201]: Either way this should be done in a separate change.

File: clang-tools-extra/clang-doc/HTMLGenerator.cpp (hunk 2 above)
Comment (PeterChou1) [reply clang-tools-extra/clang-doc/HTMLGenerator.cpp:428]: this originally meant to add support for block commands in member variables, I be happy to refactor these to its own pr though

Comment (ilovepi) [reply clang-tools-extra/clang-doc/HTMLGenerator.cpp:428]: Yeah, lets split this up a bit. I'll leave the ordering up to you, but the other feature seems to be roughly independent.

File: clang-tools-extra/unittests/clang-doc/HTMLGeneratorTest.cpp (hunk 1 above)
Comment (ilovepi) [reply clang-tools-extra/unittests/clang-doc/HTMLGeneratorTest.cpp:201]: Agreed.

File: clang-tools-extra/clang-doc/HTMLGenerator.cpp (hunk 2 above)
Comment (PeterChou1) [reply clang-tools-extra/clang-doc/HTMLGenerator.cpp]: I've split the pr to https://github.com/llvm/llvm-project/pull/101255

File: clang-tools-extra/unittests/clang-doc/HTMLGeneratorTest.cpp (hunk 1 above)
Comment (PeterChou1) [reply clang-tools-extra/unittests/clang-doc/HTMLGeneratorTest.cpp]: currently there's no distinction between public protected and private that's something i plan to change but not in this pr

File: clang-tools-extra/clang-doc/HTMLGenerator.cpp [hunk 3]
```diff
@@ -631,7 +631,19 @@ static std::unique_ptr<HTMLNode> genHTML(const CommentInfo &I) {
       return nullptr;
     return std::move(ParagraphComment);
   }
-
```
Comment (ilovepi) [inline clang-tools-extra/clang-doc/HTMLGenerator.cpp:634]: Why remove the new line?

File: clang-tools-extra/clang-doc/HTMLGenerator.cpp [hunk 4]
```diff
@@ -619,7 +620,6 @@ static std::unique_ptr<HTMLNode> genHTML(const CommentInfo &I) {
     }
     return std::move(FullComment);
   }
-
```
Comment (ilovepi) [inline clang-tools-extra/clang-doc/HTMLGenerator.cpp:622]: Unrelated?

File: clang-tools-extra/clang-doc/HTMLGenerator.cpp [hunk 5]
```diff
@@ -632,6 +633,19 @@ static std::unique_ptr<HTMLNode> genHTML(const CommentInfo &I) {
     return std::move(ParagraphComment);
   }
 
+  if (I.Kind == "BlockCommandComment") {
+    auto BlockComment = std::make_unique<TagNode>(HTMLTag::TAG_DIV);
+    auto Command = std::make_unique<TagNode>(HTMLTag::TAG_DIV, I.Name);
+    BlockComment->Children.emplace_back(std::move(Command));
```
Comment (ilovepi) [inline clang-tools-extra/clang-doc/HTMLGenerator.cpp:639]: You can avoid the move here and just construct in place, right?

File: clang-tools-extra/clang-doc/HTMLGenerator.cpp [hunk 6]
```diff
@@ -632,6 +633,19 @@ static std::unique_ptr<HTMLNode> genHTML(const CommentInfo &I) {
...
+    for (const auto &Child : I.Children) {
+      std::unique_ptr<HTMLNode> Node = genHTML(*Child);
+      if (Node)
+        BlockComment->Children.emplace_back(std::move(Node));
+    }
+    if (BlockComment->Children.empty())
+      return nullptr;
+    return std::move(BlockComment);
```
Comment (ilovepi) [inline clang-tools-extra/clang-doc/HTMLGenerator.cpp:647]: Do these need to be returned via std::move? Iirc the compiler should do the right thing with return values…

The pattern exists elsewhere in this file, it’s not required to change now, but we may want to consider up’s this code to use a more idiomatic style.

File: clang-tools-extra/test/clang-doc/basic-project.test [hunk 7]
```diff
@@ -59,32 +59,63 @@
...
-// HTML-CALC:  <p>public int subtract(int a, int b)</p>
-// HTML-CALC:  <p>Defined at line 7 of file {{.*}}Calculator.cpp</p>
-// HTML-CALC:  <h3 id="{{([0-9A-F]{40})}}">multiply</h3>
-// HTML-CALC:  <p>public int multiply(int a, int b)</p>
-// HTML-CALC:  <p>Defined at line 11 of file {{.*}}Calculator.cpp</p>
-// HTML-CALC:  <h3 id="{{([0-9A-F]{40})}}">divide</h3>
-// HTML-CALC:  <p>public double divide(int a, int b)</p>
-// HTML-CALC:  <p>Defined at line 15 of file {{.*}}Calculator.cpp</p>
+
+// HTML-CALC: <h1>class Calculator</h1>
```
Comment (ilovepi) [inline clang-tools-extra/test/clang-doc/basic-project.test:83]: Not your problem per de, but why is this line in the diff? AFACT it’s unchanged…

File: clang-tools-extra/clang-doc/HTMLGenerator.cpp (hunk 6 above)
Comment (PeterChou1) [reply clang-tools-extra/clang-doc/HTMLGenerator.cpp:647]: I could add another pr once this is merged to refactor some of this

File: clang-tools-extra/test/clang-doc/basic-project.test (hunk 7 above)
Comment (PeterChou1) [reply clang-tools-extra/test/clang-doc/basic-project.test:83]: I'm not sure, I'm not too familiar with what git considers what a diff is

File: clang-tools-extra/clang-doc/HTMLGenerator.cpp [hunk 8]
```diff
@@ -632,6 +633,19 @@ static std::unique_ptr<HTMLNode> genHTML(const CommentInfo &I) {
...
   }
 
+  if (I.Kind == "BlockCommandComment") {
+    auto BlockComment = std::make_unique<TagNode>(HTMLTag::TAG_DIV);
+    BlockComment->Children.emplace_back(
+        std::make_unique<TagNode>(HTMLTag::TAG_DIV, I.Name));
+    for (const auto &Child : I.Children) {
+      std::unique_ptr<HTMLNode> Node = genHTML(*Child);
+      if (Node)
+        BlockComment->Children.emplace_back(std::move(Node));
```
Comment (ilovepi) [inline clang-tools-extra/clang-doc/HTMLGenerator.cpp:643]:
```suggestion
      if(std::unique_ptr<HTMLNode> Node = genHTML(*Child))
        BlockComment->Children.emplace_back(std::move(Node));
```

File: clang-tools-extra/clang-doc/HTMLGenerator.cpp (hunk 6 above)
Comment (ilovepi) [reply clang-tools-extra/clang-doc/HTMLGenerator.cpp:647]: yeah, lets follow this up in another patch.

File: clang-tools-extra/test/clang-doc/basic-project.test [hunk 9]
```diff
@@ -59,32 +59,63 @@
...
-// HTML-CALC:  <h3 id="{{([0-9A-F]{40})}}">subtract</h3>
-// HTML-CALC:  <p>public int subtract(int a, int b)</p>
-// HTML-CALC:  <p>Defined at line 7 of file {{.*}}Calculator.cpp</p>
-// HTML-CALC:  <h3 id="{{([0-9A-F]{40})}}">multiply</h3>
-// HTML-CALC:  <p>public int multiply(int a, int b)</p>
-// HTML-CALC:  <p>Defined at line 11 of file {{.*}}Calculator.cpp</p>
-// HTML-CALC:  <h3 id="{{([0-9A-F]{40})}}">divide</h3>
-// HTML-CALC:  <p>public double divide(int a, int b)</p>
-// HTML-CALC:  <p>Defined at line 15 of file {{.*}}Calculator.cpp</p>
+
```
Comment (ilovepi) [inline clang-tools-extra/test/clang-doc/basic-project.test:82]: nit: extra newline

File: clang-tools-extra/test/clang-doc/basic-project.test (hunk 7 above)
Comment (ilovepi) [reply clang-tools-extra/test/clang-doc/basic-project.test:83]: I have a feeling its the extra newline.

**Review by ilovepi — APPROVED** [review]
LGTM. I left a few minor comments. Feel free to land once those are done and CI is passing.

File: clang-tools-extra/clang-doc/HTMLGenerator.cpp (hunk 8 above)
Comment (ilovepi) [reply clang-tools-extra/clang-doc/HTMLGenerator.cpp:643]: oh, you may also be able to remove the surrounding braces, I think.

**Pull request merged**